File: examples/tests/test_stats.go
// ============================================================================
// test_stats.go — StatsCore 单元测试（Go 示例）
// 对应模板：templates/NewModule/{Name}Core_test.go.template
// 运行：cd examples && go test ./tests/... -v
// ============================================================================
package tests

import (
	"math"
	"testing"

	"examples/src/stats"
)

func TestMean(t *testing.T) {
	tests := []struct {
		name  string
		input []float64
		want  float64
		isNaN bool
	}{
		{"normal positive", []float64{1, 2, 3, 4, 5}, 3.0, false},
		{"normal negative", []float64{-1, -2, -3}, -2.0, false},
		{"mixed signs", []float64{-1, 1}, 0.0, false},
		{"all zeros is valid", []float64{0, 0, 0}, 0.0, false},
		{"single element", []float64{42}, 42.0, false},
		{"nil returns NaN", nil, 0, true},
		{"empty returns NaN", []float64{}, 0, true},
		{"NaN elements filtered", []float64{1, math.NaN(), 3}, 2.0, false},
		{"Inf elements filtered", []float64{1, math.Inf(1), 3}, 2.0, false},
		{"all invalid returns NaN", []float64{math.NaN(), math.Inf(1)}, 0, true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := stats.Mean(tt.input)
			if tt.isNaN {
				if !math.IsNaN(got) {
					t.Errorf("Mean() = %v, want NaN", got)
				}
			} else if got != tt.want {
				t.Errorf("Mean() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestWeightedMean(t *testing.T) {
	tests := []struct {
		name    string
		values  []float64
		weights []float64
		want    float64
		isNaN   bool
	}{
		{"equal weights", []float64{1, 2, 3}, []float64{1, 1, 1}, 2.0, false},
		{"weighted", []float64{1, 2, 3}, []float64{0, 0, 1}, 3.0, false},
		{"nil values", nil, []float64{1, 2}, 0, true},
		{"nil weights", []float64{1, 2}, nil, 0, true},
		{"length mismatch", []float64{1, 2}, []float64{1}, 0, true},
		{"zero total weight", []float64{1, 2}, []float64{0, 0}, 0, true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := stats.WeightedMean(tt.values, tt.weights)
			if tt.isNaN {
				if !math.IsNaN(got) {
					t.Errorf("WeightedMean() = %v, want NaN", got)
				}
			} else if got != tt.want {
				t.Errorf("WeightedMean() = %v, want %v", got, tt.want)
			}
		})
	}
}

File: examples/tests/test_stats.rs
// ============================================================================
// test_stats.rs — StatsCore 集成测试（Rust 示例）
// 对应模板：templates/NewModule/{Name}Core.rs.template（#[cfg(test)] 内联测试）
// 运行：cd examples && cargo test
// ============================================================================

use examples::stats::{mean, weighted_mean};

#[test]
fn test_mean() {
    let cases: [(&str, &[f64], f64, bool); 10] = [
        ("normal positive", &[1.0, 2.0, 3.0, 4.0, 5.0], 3.0, false),
        ("normal negative", &[-1.0, -2.0, -3.0], -2.0, false),
        ("mixed signs", &[-1.0, 1.0], 0.0, false),
        ("all zeros is valid", &[0.0, 0.0, 0.0], 0.0, false),
        ("single element", &[42.0], 42.0, false),
        ("empty returns NaN", &[], 0.0, true),
        ("NaN elements filtered", &[1.0, f64::NAN, 3.0], 2.0, false),
        ("Inf elements filtered", &[1.0, f64::INFINITY, 3.0], 2.0, false),
        ("all invalid returns NaN", &[f64::NAN, f64::INFINITY], 0.0, true),
        ("overflow returns NaN", &[1e308, 1e308], 0.0, true),
    ];
    for (name, input, want, want_nan) in cases {
        let got = mean(input);
        if want_nan {
            assert!(got.is_nan(), "{name}: mean() = {got}, want NaN");
        } else {
            assert_eq!(got, want, "{name}");
        }
    }
}

#[test]
fn test_weighted_mean() {
    let cases: [(&str, &[f64], &[f64], f64, bool); 6] = [
        ("equal weights", &[1.0, 2.0, 3.0], &[1.0, 1.0, 1.0], 2.0, false),
        ("weighted", &[1.0, 2.0, 3.0], &[0.0, 0.0, 1.0], 3.0, false),
        ("empty values", &[], &[1.0, 2.0], 0.0, true),
        ("empty weights", &[1.0, 2.0], &[], 0.0, true),
        ("length mismatch", &[1.0, 2.0], &[1.0], 0.0, true),
        ("zero total weight", &[1.0, 2.0], &[0.0, 0.0], 0.0, true),
    ];
    for (name, values, weights, want, want_nan) in cases {
        let got = weighted_mean(values, weights);
        if want_nan {
            assert!(got.is_nan(), "{name}: weighted_mean() = {got}, want NaN");
        } else {
            assert_eq!(got, want, "{name}");
        }
    }
}
